staging/src/k8s.io/apiextensions-apiserver/test/integration/basic_test.go

```diff
@@ -197,6 +199,149 @@ func testSimpleCRUD(t *testing.T, ns string, noxuDefinition *apiextensionsv1beta
+	noxuResourceClient := NewNamespacedCustomResourceClient(ns, noxuVersionClient, noxuDefinition)
+	initialList, err := noxuResourceClient.List(metav1.ListOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if e, a := 0, len(initialList.(*unstructured.UnstructuredList).Items); e != a {
+		t.Errorf("expected %v, got %v", e, a)
+	}
+	initialListTypeMeta, err := meta.TypeAccessor(initialList)
+	if err != nil {
+		t.Fatal(err)
+	}
+	if e, a := noxuDefinition.Spec.Group+"/"+noxuDefinition.Spec.Version, initialListTypeMeta.GetAPIVersion(); e != a {
+		t.Errorf("expected %v, got %v", e, a)
+	}
+	if e, a := noxuDefinition.Spec.Names.ListKind, initialListTypeMeta.GetKind(); e != a {
+		t.Errorf("expected %v, got %v", e, a)
+	}
+
+	initialListListMeta, err := meta.ListAccessor(initialList)
+	if err != nil {
+		t.Fatal(err)
+	}
+	noxuWatch, err := noxuResourceClient.Watch(
+		metav1.ListOptions{
+			ResourceVersion: initialListListMeta.GetResourceVersion(),
+			FieldSelector:   "metadata.name=foo",
+		},
+	)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer noxuWatch.Stop()
+
+	_, err = instantiateCustomResource(t, testserver.NewNoxuInstance(ns, "bar"), noxuResourceClient, noxuDefinition)
+	if err != nil {
+		t.Fatalf("unable to create noxu Instance:%v", err)
+	}
+	createdNoxuInstanceFoo, err := instantiateCustomResource(t, testserver.NewNoxuInstance(ns, "foo"), noxuResourceClient, noxuDefinition)
+	if err != nil {
+		t.Fatalf("unable to create noxu Instance:%v", err)
+	}
+
+	select {
+	case watchEvent := <-noxuWatch.ResultChan():
+		if e, a := watch.Added, watchEvent.Type; e != a {
+			t.Errorf("expected %v, got %v", e, a)
+			break
+		}
+		createdObjectMeta, err := meta.Accessor(watchEvent.Object)
+		if err != nil {
+			t.Fatal(err)
+		}
+		// it should have a UUID
+		if len(createdObjectMeta.GetUID()) == 0 {
+			t.Errorf("missing uuid: %#v", watchEvent.Object)
+		}
+		if e, a := ns, createdObjectMeta.GetNamespace(); e != a {
+			t.Errorf("expected %v, got %v", e, a)
+		}
+		if e, a := "foo", createdObjectMeta.GetName(); e != a {
+			t.Errorf("expected %v, got %v", e, a)
+		}
+		createdTypeMeta, err := meta.TypeAccessor(watchEvent.Object)
+		if err != nil {
+			t.Fatal(err)
+		}
+		if e, a := noxuDefinition.Spec.Group+"/"+noxuDefinition.Spec.Version, createdTypeMeta.GetAPIVersion(); e != a {
+			t.Errorf("expected %v, got %v", e, a)
+		}
+		if e, a := noxuDefinition.Spec.Names.Kind, createdTypeMeta.GetKind(); e != a {
+			t.Errorf("expected %v, got %v", e, a)
+		}
+
+	case <-time.After(5 * time.Second):
+		t.Errorf("missing watch event")
+	}
+
+	gottenNoxuInstance, err := noxuResourceClient.Get("foo", metav1.GetOptions{})
+	if err != nil {
+		t.Fatal(err)
+	}
+	if e, a := createdNoxuInstanceFoo, gottenNoxuInstance; !reflect.DeepEqual(e, a) {
+		t.Errorf("expected %v, got %v", e, a)
+	}
+
+	listWithItem, err := noxuResourceClient.List(metav1.ListOptions{FieldSelector: "metadata.name=foo"})
```

Comment:
Could you arrange for the objects to differ in a custom spec field, and select for that here? It would be good to test that it works for fields outside metadata.

Reply:
That doesn't work. I believe it's because of https://github.com/kubernetes/kubernetes/blob/2bf7a20a542ac54ccb6c284aa0b01559bf02a43b/staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresource/strategy.go#L98-L109

Reply:
I could write some code that creates a `fields.Set` of every possible field leaf (excluding slices); e.g., if we have this:
```yaml
a: b
c: 1
d:
  - a
  - b
e:
  f: g
  h:
    i: j
```
the field set would be
```yaml
a: b
c: 1
e.f: g
e.h.i: j
```

Reply:
I think you're right. I must have misremembered that it worked for other fields in my earlier experiment. Should we specify in the release note that CRD field selectors only support name and namespace?

Reply:
Sample code that works: https://gist.github.com/ncdc/edc1ed3abefa36d9650a524ebd5f3f8c

Reply:
Would metadata.name and metadata.namespace solve the original concern that caused you to file #51046? If so, I'd rather not add the full field dump before anyone has asked for it, mainly because of the precedent set by other resources like Pod, which is very conservative in its field selector whitelist:

https://github.com/kubernetes/kubernetes/blob/659a68978265df1c48838a18c31016b9da095722/pkg/registry/core/pod/strategy.go#L240-L244

If the feature is ever requested, perhaps it would make more sense to let CRD authors decide the whitelist in the CRD spec.

Reply:
I don't 100% remember what I was doing when I filed that issue, but I'm pretty confident that being able to use metadata.name as a field selector for a custom resource would be sufficient for my immediate needs.

I also assume that the sample code in my gist could have very negative performance implications, so I'm fine not pulling it in for now.

+1 to finding a way to specify the whitelist in the CRD spec.

Reply:
I would love to be able to use arbitrary fields from a CR spec in a field selector. It's not clear to me whether there are any actual concerns/issues with allowing users to make queries using whatever field they want. Unfortunately, the metadata fields are useless in my case.

Reply:
@kargakis I don't have any concerns with the UX of allowing arbitrary fields; that was my plan when I thought all we needed to do was stop blocking it. However, it turns out we also have to recursively flatten the whole custom object into a map, for every object returned by a list, in order to make filtering work for arbitrary fields.

That seems extreme given how conservative other objects are with their supported field selectors. For example, Pod only picks out 3 fields from spec that you're allowed to use in fieldSelector.

If you have a use case for filtering on arbitrary fields, would you be ok with having to specify a whitelist in the CRD spec, so we don't have to export every field regardless of whether anyone wants it?

We could discuss that on a separate feature request issue. Either way I think we should go ahead and merge this fix for metadata.name, since that's important for watching single objects.

Reply:
Agreeing to go ahead with this PR before discussing abritrary keys in a follow-up.

Reply:
@enisoc yeah, a whitelist would be great to have. Please cc me one the new issue when one is made.

Reply:
@kargakis I think it's better if you file the issue. I don't actually have any use case in mind for needing arbitrary field selectors, so I can't explain why the feature is being requested.

Reply:
Opened https://github.com/kubernetes/kubernetes/issues/53459